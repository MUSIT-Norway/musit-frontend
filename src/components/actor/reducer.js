import { Observable } from 'rxjs/Rx';
import actions from "./actions";

const dataFromServer = [
  {
    id: 1,
    fn: 'Jarl parl'
  },
  {
    id: 2,
    fn: 'Line pine'
  },
  {
    id: 3,
    fn: 'Kjell pell'
  },
  {
    id: 4,
    fn: 'Stein sein'
  },
  {
    id: 5,
    fn: 'Knut lut'
  },
  {
    id: 6,
    fn: 'Ellen'
  }
]

const initialState = [];

const ActorReducer$ = Observable.of(() => initialState)
  .merge(
    actions.suggest$.map((update) => () => {
      if (!update) {
        return initialState;
      }
      return dataFromServer.filter(d => {
        const name = d.fn.toLowerCase();
        const query = update.value.toLowerCase();
        return name.indexOf(query) > -1
      })
    }),
    actions.clear$.map(() => () => initialState),
  );

export default Observable.merge(
  ActorReducer$.map(reducer => ["suggest", reducer]),
);