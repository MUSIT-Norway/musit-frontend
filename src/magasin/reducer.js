import { toFrontend } from './mapper';
import * as types from './types';

const initialState = {
  root: {
    data: {}
  }
};

const storageUnitGridReducer = (state = initialState, action = {}) => {
  switch (action.type) {
  case types.LOAD_SEVERAL:
    return {
      ...state,
      loading: true
    };
  case types.LOAD_SEVERAL_SUCCESS:
    return {
      ...state,
      loading: false,
      loaded: true,
      data: action.result
    };
  case types.LOAD_SEVERAL_FAIL:
    return {
      ...state,
      loading: false,
      loaded: false,
      error: action.error
    };
  case types.LOAD_ONE_SUCCESS:
    return {
      ...state,
      root: {
        ...state.root,
        loading: false,
        loaded: true,
        data: toFrontend(action.result)
      }
    };
  case types.LOAD_ONE_FAIL:
    return {
      ...state,
      root: {
        ...state.root,
        loading: false,
        loaded: false,
        error: action.error
      }
    };
  case types.CLEAR_ROOT: {
    return {
      ...state,
      root: {}
    };
  }
  case types.DELETE:
    return {
      ...state,
      loading: true
    };
  case types.DELETE_SUCCESS:
    return {
      ...state,
      loading: false,
      loaded: true,
      data: []
    };
  case types.DELETE_FAIL:
    return {
      ...state,
      loading: false,
      loaded: false,
      error: action.error
    };
  default:
    return state;
  }
};

export default storageUnitGridReducer;