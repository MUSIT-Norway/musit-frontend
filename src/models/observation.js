import entries from 'object.entries';
import Config from '../config';
import mapToBackEnd from './mapper/observation/to_backend';
import mapToFrontEnd from './mapper/observation/to_frontend';
import MusitActor from './actor';
import uniq from 'lodash/uniq';
import { simplePost, simpleGet } from '../shared/RxAjax';

class Observation {
  constructor(props) {
    entries(props).forEach(([k, v]) => this[k] = v);
  }
}

Observation.loadObservations = (ajaxGet = simpleGet) => ({ nodeId, museumId, token, callback }) => {
  return ajaxGet(`${Config.magasin.urls.storagefacility.baseUrl(museumId)}/${nodeId}/observations`, token, callback)
    .map(({ response }) => {
      if (!Array.isArray(response)) {
        return [];
      }
      return response.map(json => new Observation(json));
    });
};

Observation.addObservation = (ajaxPost = simplePost) => ({ nodeId, museumId, data, token, callback }) => {
  const url = `${Config.magasin.urls.storagefacility.baseUrl(museumId)}/${nodeId}/observations`;
  const dataToPost = mapToBackEnd(data, nodeId);
  console.log('Model observation: ', dataToPost);
  return ajaxPost(url, dataToPost, token, callback);
};

Observation.getObservation = (ajaxGet = simpleGet, ajaxPost = simplePost) => ({ nodeId, observationId, museumId, token }) => {
  const url =`${Config.magasin.urls.storagefacility.baseUrl(museumId)}/${nodeId}/observations/${observationId}`;
  return ajaxGet(url, token)
    .flatMap(observation => {
      const actorIds = uniq([observation.response.doneBy, observation.response.registeredBy]).filter(p => p);
      return MusitActor.getActors(ajaxPost)(actorIds, token)
        .map(actorDetails => {
          return new Observation(mapToFrontEnd({
            ...observation.response,
            ...MusitActor.getActorNames(actorDetails, observation.response.doneBy, observation.response.registeredBy)
          }));
        });
    });
};


export default Observation;