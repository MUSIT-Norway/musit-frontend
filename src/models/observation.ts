// @flow
import Config from '../config';
import mapToBackEnd from './mapper/observation/to_backend';
import mapToFrontEnd from './mapper/observation/to_frontend';
import MusitActor from './actor';
import { uniq } from 'lodash';
import { simplePost, simpleGet } from '../shared/RxAjax';
import { Callback, AjaxGet, AjaxPost } from '../types/ajax';
import { Observable } from 'rxjs';
import { Star, mixed } from '../types/common';

class Observation {
  static loadObservations: (
    ajaxGet: AjaxGet<Star>
  ) => (
    props: {
      nodeId: number;
      museumId: number;
      token: string;
      callback?: Callback<Star>;
    }
  ) => Observable<Star>;
  static addObservation: (
    ajaxPost?: AjaxPost<Star>
  ) => (
    props: {
      nodeId: number;
      museumId: number;
      data: mixed;
      token: string;
      callback?: Callback<Star>;
    }
  ) => Observable<Star>;
  static getObservation: (
    ajaxGet?: AjaxGet<Star>,
    ajaxPost?: AjaxPost<Star>
  ) => (
    props: {
      nodeId: number;
      observationId: number;
      museumId: number;
      token: string;
    }
  ) => Observable<Star>;
}

Observation.loadObservations = (ajaxGet = simpleGet) => ({
  nodeId,
  museumId,
  token,
  callback
}) => {
  return ajaxGet(
    `${Config.magasin.urls.api.storagefacility.baseUrl(museumId)}/${nodeId}/observations`,
    token,
    callback
  ).map(({ response }) => {
    if (!Array.isArray(response)) {
      return [];
    }
    return response;
  });
};

Observation.addObservation = (ajaxPost = simplePost) => ({
  nodeId,
  museumId,
  data,
  token,
  callback
}) => {
  const url = `${Config.magasin.urls.api.storagefacility.baseUrl(
    museumId
  )}/${nodeId}/observations`;
  const dataToPost = mapToBackEnd(data, nodeId);
  return ajaxPost(url, dataToPost, token, callback);
};

Observation.getObservation = (ajaxGet = simpleGet, ajaxPost = simplePost) => ({
  nodeId,
  observationId,
  museumId,
  token
}) => {
  const url = `${Config.magasin.urls.api.storagefacility.baseUrl(
    museumId
  )}/${nodeId}/observations/${observationId}`;
  return ajaxGet(url, token).flatMap(observation => {
    const actorIds = uniq([
      observation.response.doneBy,
      observation.response.registeredBy
    ]).filter(p => p);
    return MusitActor.getActors(ajaxPost)({ actorIds, token }).map(actorDetails => {
      return mapToFrontEnd({
        ...observation.response,
        ...MusitActor.getActorNames(
          actorDetails,
          observation.response.doneBy,
          observation.response.registeredBy
        )
      });
    });
  });
};

export default Observation;
