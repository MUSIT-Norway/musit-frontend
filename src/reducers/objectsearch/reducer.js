import { omit } from 'lodash';
import { getPath } from '../helper';
import MusitObject from '../../models/object';
import * as types from './constants';

const initialState = {
  data: {
    totalMatches: 0,
    matches: []
  },
  params: {
    currentPage: 1,
    perPage: 50
  },
  loaded: true
};

export default (state = initialState, action) => {
  switch(action.type) {
  case types.CHANGE_FIELD:
    return {
      ...state,
      params: {
        ...state.params,
        [action.field]: action.value
      }
    };
  case types.SEARCH_OBJECTS:
    return {
      ...omit(state, 'error'),
      loading: true,
      loaded: false,
      data: {
        totalMatches: 0,
        matches: []
      },
      params: {
        ...state.params,
        currentPage: action.page
      }
    };
  case types.SEARCH_OBJECTS_FAIL:
    return {
      ...state,
      loading: false,
      loaded: false,
      data: [],
      error: action.error
    };
  case types.SEARCH_OBJECTS_SUCCESS:
    return {
      ...omit(state, 'error'),
      loading: false,
      loaded: true,
      data: {
        totalMatches: action.result.totalMatches,
        matches: action.result.matches.map(data => {
          return new MusitObject({
            ...data,
            breadcrumb: getPath(data)
          });
        })
      }
    };
  default:
    return state;
  }
};