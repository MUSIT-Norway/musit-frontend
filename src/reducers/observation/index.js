import Config from '../../config';
import { apiUrl } from '../../util';
import mapToBackEnd from './mapper/to_backend';
import mapToFrontEnd from './mapper/to_frontend';
import uniq from 'lodash/uniq';
import Actor from '../../models/actor';

export const ADD = 'musit/observation/ADD';
export const ADD_SUCCESS = 'musit/observation/ADD_SUCCESS';
export const ADD_FAIL = 'musit/observation/ADD_FAIL';
export const LOAD = 'musit/observation/LOAD';
export const LOAD_SUCCESS = 'musit/observation/LOAD_SUCCESS';
export const LOAD_FAIL = 'musit/observation/LOAD_FAIL';
export const initialState = {
  data: {
    observations: []
  }
};

const observationReducer = (state = initialState, action = {}) => {
  let d = {};
  switch (action.type) {
  case ADD:
    return {
      ...state,
      loading: true,
      loaded: false,
      data: {}
    };
  case ADD_SUCCESS:
    d = mapToFrontEnd(action.result);
    return {
      ...state,
      loading: false,
      loaded: true,
      data: d
    };
  case ADD_FAIL:
    return {
      ...state,
      loading: false,
      loaded: false,
      data: action.error
    };
  case LOAD:
    return {
      ...state,
      loading: true,
      loaded: false,
      data: {}
    };
  case LOAD_SUCCESS:
    d = mapToFrontEnd(action.result);
    return {
      ...state,
      loading: false,
      loaded: true,
      data: d
    };
  case LOAD_FAIL:
    return {
      ...state,
      loading: false,
      loaded: false,
      data: action.error
    };
  default:
    return state;
  }
};

export default observationReducer;

export const addObservation = (nodeId, data, callback) => {
  const action = 'post';
  const url = apiUrl(`${Config.magasin.urls.storagefacility.baseUrl(99)}/${nodeId}/observations`);
  const dataToPost = mapToBackEnd(data, nodeId);
  return {
    types: [ADD, ADD_SUCCESS, ADD_FAIL],
    promise: (client) => client[action](url, { data: dataToPost }),
    callback
  };
};

export const loadObservation = (nodeId, observationId, callback) => {
  const url = apiUrl(`${Config.magasin.urls.storagefacility.baseUrl(99)}/${nodeId}/observations/${observationId}`);
  return {
    types: [LOAD, LOAD_SUCCESS, LOAD_FAIL],
    promise: (client) => new Promise((resolve, reject) => {
      client.get(url).then(observation => {
        client.post(apiUrl(`${Config.magasin.urls.actor.baseUrl}/details`), { data: uniq([observation.doneBy, observation.registeredBy]) })
            .then(actors =>
              resolve({
                ...observation,
                ...Actor.getActorNames(actors, observation.doneBy, observation.registeredBy)
              })
            ).catch(error => reject(error));
      }).catch(error => reject(error));
    }),
    callback
  };
};
