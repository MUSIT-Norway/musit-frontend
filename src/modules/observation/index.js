import reducer from './observationReducers';
import * as actions from './observationReducers';
import routes from './observationRoutes';

const api = {
  reducer,
  actions,
  routes
};

export default api;