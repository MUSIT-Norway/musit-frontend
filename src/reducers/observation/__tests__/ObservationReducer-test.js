import assert from 'assert';
import observationReducer, {
  loadObservation,
  initialState
} from '../index';
import mapToFrontEnd from '../mapper/to_frontend';
import mapToBackEnd from '../mapper/to_backend';
import Actor from '../../../models/actor';
import configureMockStore from 'redux-mock-store';
import createMiddleware from '../../../middleware/clientMiddleware';
import ApiClient from '../../../middleware/ApiClient';
import Config from '../../../config';
import request from 'superagent';
import nocker from 'superagent-nock';
import reducer from '../../../reducers/observation/index';
import * as actions from '../../../reducers/observation/index';
import MuseumId from '../../../models/museumId';

const nock = nocker(request);
const middlewares = [createMiddleware(new ApiClient())];
const mockStore = configureMockStore(middlewares);


describe('ObservationReducer', () => {
  it('Initial state is set', () => {
    const state = observationReducer();
    assert(state === initialState);
  });

  it('load observation should update state', () => {
    const state = observationReducer(initialState, loadObservation(1, 2, new MuseumId(99)));
    assert(state.data.observations.length === 0);
  });

  it('mapToFrontEnd and mapToBackEnd shoud be inverse functions', () => {
    const frontEnd = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {
          type: 'hypoxicAir',
          data: {
            fromValue: '19',
            toValue: '23',
            commentValue: 'Test comments.'
          }
        },
        {
          type: 'lightCondition',
          data: {
            leftValue: 'Light',
            rightValue: 'Test comments.'
          }
        },
        {
          type: 'cleaning',
          data: {
            leftValue: 'Test cleaning value.',
            rightValue: 'Test comments.'
          }
        }
      ]
    };
    const backEnd = mapToBackEnd(frontEnd);
    expect(backEnd).toMatchSnapshot();
    const thereAndBackAgain = mapToFrontEnd(backEnd);
    expect(thereAndBackAgain).toMatchSnapshot();
  });

  it('Valid action from calling actor service for finding actor should update state correctly', () => {
    const actionResult = {
      data: {
        id: 1,
        fn: 'Arne And1'
      },
      type: 'LOAD_ACTOR_SUCCESS'
    };

    const state = observationReducer(initialState, actionResult);
    assert(state.type !== 'LOAD_ACTOR_SUCCESS' || state.data.fn === 'Arne And1');
  });

  it('test alcohol status: Uttørket', () => {
    const frontend = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {

          type: 'alcohol',
          data: {
            statusValue: 'Uttørket',
            volumeValue: '4',
            commentValue: 'Drita full'
          }
        }
      ]
    };
    const backEnd = mapToBackEnd(frontend);
    expect(backEnd).toMatchSnapshot();
    const thereAndBackAgain = mapToFrontEnd(backEnd);
    expect(thereAndBackAgain).toMatchSnapshot();
  });

  it('test alcohol status: nesten uttørket', () => {
    const frontend = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {

          type: 'alcohol',
          data: {
            statusValue: 'nesten uttørket',
            volumeValue: '4',
            commentValue: 'Drita full'
          }
        }
      ]
    };
    const backEnd = mapToBackEnd(frontend);
    expect(backEnd).toMatchSnapshot();
    const thereAndBackAgain = mapToFrontEnd(backEnd);
    expect(thereAndBackAgain).toMatchSnapshot();
  });

  it('test alcohol status: litt uttørket', () => {
    const frontend = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {

          type: 'alcohol',
          data: {
            statusValue: 'litt uttørket',
            volumeValue: '4',
            commentValue: 'Drita full'
          }
        }
      ]
    };
    const backEnd = mapToBackEnd(frontend);
    expect(backEnd).toMatchSnapshot();
    const thereAndBackAgain = mapToFrontEnd(backEnd);
    expect(thereAndBackAgain).toMatchSnapshot();
  });

  it('test alcohol status: noe uttørket', () => {
    const frontend = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {

          type: 'alcohol',
          data: {
            statusValue: 'noe uttørket',
            volumeValue: '4',
            commentValue: 'Drita full'
          }
        }
      ]
    };
    const backEnd = mapToBackEnd(frontend);
    expect(backEnd).toMatchSnapshot();
    const thereAndBackAgain = mapToFrontEnd(backEnd);
    expect(thereAndBackAgain).toMatchSnapshot();
  });

  it('test alcohol status: tilfredsstillende', () => {
    const frontend = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {

          type: 'alcohol',
          data: {
            statusValue: 'tilfredsstillende',
            volumeValue: '4',
            commentValue: 'Drita full'
          }
        }
      ]
    };
    const backEnd = mapToBackEnd(frontend);
    expect(backEnd).toMatchSnapshot();
    const thereAndBackAgain = mapToFrontEnd(backEnd);
    expect(thereAndBackAgain).toMatchSnapshot();

  });

  it('test invalid alcohol status', () => {
    const frontend = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {

          type: 'alcohol',
          data: {
            statusValue: 'xxxddudyudydydydydyd',
            volumeValue: '4',
            commentValue: 'Drita full'
          }
        }
      ]
    };
    const backEnd = mapToBackEnd(frontend);
    expect(backEnd).toMatchSnapshot();
    const thereAndBackAgain = mapToFrontEnd(backEnd);
    expect(thereAndBackAgain).toMatchSnapshot();
  });

  it('mapToFrontEnd and mapToBackEnd are inverse with complete data', () => {
    const frontend = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {

          type: 'lightCondition',
          data: {
            leftValue: 'Mørkst',
            rightValue: 'Altfor mørkt'
          }
        },
        {

          type: 'gas',
          data: {
            leftValue: 'Vannskade',
            rightValue: 'Altfor vått'
          }
        },
        {

          type: 'alcohol',
          data: {
            statusValue: 'Uttørket',
            volumeValue: '4',
            commentValue: 'Drita full'
          }
        },
        {

          type: 'mold',
          data: {
            leftValue: 'Muggent',
            rightValue: 'Altfor mye mugg'
          }
        },
        {

          type: 'cleaning',
          data: {
            leftValue: 'Urent',
            rightValue: 'Altfor lyst'
          }
        },
        {

          type: 'brannsikring',
          data: {
            leftValue: 'Brann',
            rightValue: 'Altfor vått'
          }
        },
        {

          type: 'skallsikring',
          data: {
            leftValue: 'skall',
            rightValue: 'Altfor skallet'
          }
        },
        {

          type: 'tyverisikring',
          data: {
            leftValue: 'Mye tyver',
            rightValue: 'Altfor mye tyver'
          }
        },
        {

          type: 'hypoxicAir',
          data: {
            fromValue: '1,4',
            toValue: '4,4',
            commentValue: 'Altfor fuktig'
          }
        },
        {

          type: 'relativeHumidity',
          data: {
            fromValue: '1,4',
            toValue: '4,4',
            commentValue: 'Altfor fuktig'
          }
        },
        {
          type: 'pest',
          data: {
            identificationValue: 'Mye skadedyr',
            commentValue: 'Ikke gjort noe med',
            observations: [
              {
                lifeCycle: 'Larva',
                count: '4'
              },
              {
                lifeCycle: 'Puppe',
                count: '32'
              }
            ]
          }
        },
        {
          type: 'vannskaderisiko',
          data: {
            leftValue: 'Vannskade',
            rightValue: 'Altfor vått'
          }
        },
        {
          type: 'temperature',
          data: {
            fromValue: '1,4',
            toValue: '4,4',
            commentValue: 'Altfor fuktig'
          }
        }
      ]
    };
    const backEnd = mapToBackEnd(frontend);
    expect(backEnd).toMatchSnapshot();
    const thereAndBackAgain = mapToFrontEnd(backEnd);
    expect(thereAndBackAgain).toMatchSnapshot();
  });


  it('creates LOAD_SUCCESS when observation data is loaded', () => {
    const nodeId = 2;
    const observationId = 3;
    const url = `${Config.magasin.urls.storagefacility.baseUrl(new MuseumId(99))}/${nodeId}/observations/${observationId}`;
    nock('http://localhost')
      .get(url)
      .reply(200, {
        id: 50,
        doneBy: '8994945c-89a1-4086-b17a-728df8a907a4',
        doneDate: '2016-09-29T00:00:00+00:00',
        affectedThing: 2,
        registeredBy: 'Darth Vader',
        registeredDate: '2016-10-17T14:22:39+00:00',
        eventType: 'Observation',
        gas: {
          gas: 'Gas test'
        }
      });

    const actorUrl = `${Config.magasin.urls.actor.baseUrl}/details`;
    nock('http://localhost')
      .post(actorUrl, ['8994945c-89a1-4086-b17a-728df8a907a4'])
      .reply(200, [{ fn: 'Test user', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4'}]);

    const store = mockStore();

    return store.dispatch(actions.loadObservation(2, 3, new MuseumId(99)))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });

  it('Observation no action', () => {
    expect(
      reducer(undefined, {})
    ).toMatchSnapshot();
  });

  it('Observation initial action', () => {
    expect(
      reducer(undefined, {
        type: actions.LOAD
      })
    ).toMatchSnapshot();
  });

  it('Observation success action', () => {
    expect(
      reducer(undefined, {
        type: actions.LOAD_SUCCESS,
        result: {
          doneBy: '7da64e83-7918-4cbb-8d81-f71e3ecdae67'
        }
      })
    ).toMatchSnapshot();
  });

  it('Observation fail action', () => {
    expect(
      reducer(undefined, {
        type: actions.LOAD_FAIL,
        error: Error('Some error occurred to load observation data.')
      })
    ).toMatchSnapshot();
  });

  it('creates ADD_SUCCESS when observation data is added', () => {
    const observationAddData = {
      doneBy: new Actor({ id: '1', dataportenId: '8994945c-89a1-4086-b17a-728df8a907a4' }),
      doneDate: '2016-10-31T23:00:00.000Z',
      observations: [
        {
          type: 'hypoxicAir',
          data: {
            fromValue: '19',
            toValue: '23',
            commentValue: 'Test comments.'
          }
        },
        {
          type: 'cleaning',
          data: {
            leftValue: 'Test cleaning value.',
            rightValue: 'Test comments.'
          }
        }
      ]
    };
    const nodeId = 4;
    const url = `${Config.magasin.urls.storagefacility.baseUrl(new MuseumId(99))}/${nodeId}/observations`;
    nock('http://localhost')
      .post(url, observationAddData)
      .reply(201, {
        id: 10,
        doneBy: 1,
        doneDate: '2016-10-18T00:00:00+00:00',
        affectedThing: 4,
        registeredBy: 'Darth Vader',
        registeredDate: '2016-10-18T13:42:10+00:00',
        eventType: 'Observation',
        gas: {
          note: 'Gas comments',
          gas: 'test Gas'
        }
      });

    const observationStore = mockStore();

    return observationStore.dispatch(actions.addObservation(4, new MuseumId(99), observationAddData))
      .then(() => {
        expect(observationStore.getActions()).toMatchSnapshot();
      });
  });

  it('Add observation no action', () => {
    expect(
      reducer(undefined, {})
    ).toMatchSnapshot();
  });

  it('Add observation initial action', () => {
    expect(
      reducer(undefined, {
        type: actions.ADD
      })
    ).toMatchSnapshot();
  });

  it('Add observation success action', () => {
    expect(
      reducer(undefined, {
        type: actions.ADD_SUCCESS,
        result: {
          doneBy: '062f6a86-7f54-4d0b-9cbe-2b83b6af9b83'
        }
      })
    ).toMatchSnapshot();
  });

  it('Add observation fail action', () => {
    expect(
      reducer(undefined, {
        type: actions.ADD_FAIL,
        error: Error('Some error occurred in the add observation data method.')
      })
    ).toMatchSnapshot();
  });
});
