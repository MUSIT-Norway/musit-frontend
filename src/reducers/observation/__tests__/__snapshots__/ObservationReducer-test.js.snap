exports[`ObservationReducer Actor fail action 1`] = `
Object {
  "data": Object {
    "doneBy": [Error: Some error occurred to load actor data.],
    "observations": Array []
  },
  "loaded": false,
  "loading": false
}
`;

exports[`ObservationReducer Actor initial action 1`] = `
Object {
  "data": Object {
    "observations": Array []
  },
  "loaded": false,
  "loading": true
}
`;

exports[`ObservationReducer Actor no action 1`] = `
Object {
  "data": Object {
    "observations": Array []
  }
}
`;

exports[`ObservationReducer Actor success action 1`] = `
Object {
  "data": Object {
    "doneBy": Object {
      "someField": 1
    },
    "observations": Array []
  },
  "loaded": true,
  "loading": false
}
`;

exports[`ObservationReducer Add observation fail action 1`] = `
Object {
  "data": [Error: Some error occurred in the add observation data method.],
  "loaded": false,
  "loading": false
}
`;

exports[`ObservationReducer Add observation initial action 1`] = `
Object {
  "data": Object {},
  "loaded": false,
  "loading": true
}
`;

exports[`ObservationReducer Add observation no action 1`] = `
Object {
  "data": Object {
    "observations": Array []
  }
}
`;

exports[`ObservationReducer Add observation success action 1`] = `
Object {
  "data": Object {
    "doneBy": Object {
      "id": undefined
    },
    "doneDate": undefined,
    "observations": Array [],
    "registeredBy": undefined,
    "registeredDate": undefined
  },
  "loaded": true,
  "loading": false
}
`;

exports[`ObservationReducer Observation fail action 1`] = `
Object {
  "data": [Error: Some error occurred to load observation data.],
  "loaded": false,
  "loading": false
}
`;

exports[`ObservationReducer Observation initial action 1`] = `
Object {
  "data": Object {},
  "loaded": false,
  "loading": true
}
`;

exports[`ObservationReducer Observation no action 1`] = `
Object {
  "data": Object {
    "observations": Array []
  }
}
`;

exports[`ObservationReducer Observation success action 1`] = `
Object {
  "data": Object {
    "doneBy": Object {
      "id": undefined
    },
    "doneDate": undefined,
    "observations": Array [],
    "registeredBy": undefined,
    "registeredDate": undefined
  },
  "loaded": true,
  "loading": false
}
`;

exports[`ObservationReducer creates ADD_SUCCESS when observation data is added 1`] = `
Array [
  Object {
    "type": "musit/observation/ADD"
  },
  Object {
    "result": Object {
      "affectedThing": 4,
      "doneBy": 1,
      "doneDate": "2016-10-18T00:00:00+00:00",
      "eventType": "Observation",
      "gas": Object {
        "gas": "test Gas",
        "note": "Gas comments"
      },
      "id": 10,
      "registeredBy": "Darth Vader",
      "registeredDate": "2016-10-18T13:42:10+00:00"
    },
    "type": "musit/observation/ADD_SUCCESS"
  }
]
`;

exports[`ObservationReducer creates LOAD_ACTOR_SUCCESS when actor data is imported 1`] = `
Array [
  Object {
    "type": "musit/observation/actor/LOAD"
  },
  Object {
    "result": Object {
      "dataportenId": "jarle",
      "fn": "Jarle Stabell",
      "id": 1
    },
    "type": "musit/observation/actor/LOAD_SUCCESS"
  }
]
`;

exports[`ObservationReducer creates LOAD_SUCCESS when observation data is loaded 1`] = `
Array [
  Object {
    "type": "musit/observation/LOAD"
  },
  Object {
    "result": Object {
      "affectedThing": 2,
      "doneBy": 1,
      "doneDate": "2016-09-29T00:00:00+00:00",
      "eventType": "Observation",
      "gas": Object {
        "gas": "Gas test"
      },
      "id": 50,
      "registeredBy": "Darth Vader",
      "registeredDate": "2016-10-17T14:22:39+00:00"
    },
    "type": "musit/observation/LOAD_SUCCESS"
  }
]
`;
